File: PYTHON/searching_algo/binary_search.py
# def binarySearch(nums, num):
#     low = 0
#     high = len(nums)-1
#     i = 0
#     while i < len(nums)// 2:
#         mid = (low + high) //2
#         if nums[mid] == num:
#             return True
#         elif num < nums[mid]:
#             high = mid-1
#         else:
#             low = mid+1
#         i+=1
#     return False

def binarySearchRecursion(nums, num, low, high):
    if low > high:
        return False
    else:
        mid = (low + high)//2
        if nums[mid] == num:
            return True
        elif num < nums[mid]:
            return binarySearchRecursion(nums, num, low, mid-1)
        else:
            return binarySearchRecursion(nums, num, mid+1, high)

print(binarySearchRecursion([1, 2, 3, 4, 5, 6, 7, 8, 9], 10, 0, 8))

File: CPP/Searching/binarySearch.cpp
#include<iostream>
#include<stdio.h>

using namespace std;

void binarySearch(int arr[], int low, int high, int num){
    int mid = (low+high)/2;
    if(arr[mid] == num){
        cout<<num<<" is found at index "<<mid<<endl;
        return;
    }
    if(arr[mid] < num){
        binarySearch(arr, mid+1, high, num);
    }
    else if(arr[mid] > num){
        binarySearch(arr, 0, mid-1, num);
    }
}

int main()
{
    int arr[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 12, 14, 15, 18};
    int len = sizeof(arr)/sizeof(arr[0]);
    binarySearch(arr, 0, len-1, 18);
    return 0;
}
